From ed0300c50cdcf39b83c14d01257202c9681d8018 Mon Sep 17 00:00:00 2001
From: Yi-Chia Hsieh <Yi-Chia.Hsieh@mediatek.com>
Date: Fri, 15 Aug 2025 15:56:20 -0700
Subject: [PATCH 118/128] mtk: mac80211: Support for receiving BTWT s1g action
 frame

The original flow can only handle individual TWT. Extend the existing flow
for processing broadcast TWT. A broadcast TWT scheduled STA will set the
broadcast bit in control field in TWT setup frame. The process flow for
individual and broadcast TWT share the same hook functions. The underlay
driver functions can differentiate by the broadcast bit in control field.

Signed-off-by: Yi-Chia Hsieh <Yi-Chia.Hsieh@mediatek.com>
---
 include/linux/ieee80211-he.h | 11 +++++++++++
 include/net/mac80211.h       | 12 ++++++------
 net/mac80211/driver-ops.h    | 37 +++++++++++++++++++++---------------
 net/mac80211/s1g.c           | 30 ++++++++++++++---------------
 4 files changed, 54 insertions(+), 36 deletions(-)

diff --git a/include/linux/ieee80211-he.h b/include/linux/ieee80211-he.h
index 879da4eb..e4084f2c 100644
--- a/include/linux/ieee80211-he.h
+++ b/include/linux/ieee80211-he.h
@@ -24,6 +24,11 @@
 #define IEEE80211_TWT_CONTROL_RX_DISABLED		BIT(4)
 #define IEEE80211_TWT_CONTROL_WAKE_DUR_UNIT		BIT(5)
 
+#define IEEE80211_TWT_NEGO_TYPE_ITWT			0
+#define IEEE80211_TWT_NEGO_TYPE_NEXT_WAKE_TBTT		1
+#define IEEE80211_TWT_NEGO_TYPE_BTWT_ANNOUNCE		2
+#define IEEE80211_TWT_NEGO_TYPE_BTWT_MBR_MGMT		3
+
 #define IEEE80211_TWT_REQTYPE_REQUEST			BIT(0)
 #define IEEE80211_TWT_REQTYPE_SETUP_CMD			GENMASK(3, 1)
 #define IEEE80211_TWT_REQTYPE_TRIGGER			BIT(4)
@@ -33,6 +38,12 @@
 #define IEEE80211_TWT_REQTYPE_WAKE_INT_EXP		GENMASK(14, 10)
 #define IEEE80211_TWT_REQTYPE_PROTECTION		BIT(15)
 
+#define IEEE80211_TWT_REQTYPE_LAST_BCAST_PARAM		BIT(5)
+#define IEEE80211_TWT_REQTYPE_RECM			GENMASK(9, 7)
+
+#define IEEE80211_BTWT_INFO_ID				GENMASK(7, 3)
+#define IEEE80211_BTWT_INFO_PERSISTENCE			GENMASK(15, 8)
+
 enum ieee80211_twt_setup_cmd {
 	TWT_SETUP_CMD_REQUEST,
 	TWT_SETUP_CMD_SUGGEST,
diff --git a/include/net/mac80211.h b/include/net/mac80211.h
index 44937671..03b7aa4e 100644
--- a/include/net/mac80211.h
+++ b/include/net/mac80211.h
@@ -4925,12 +4925,12 @@ struct ieee80211_ops {
 	void (*sta_set_decap_offload)(struct ieee80211_hw *hw,
 				      struct ieee80211_vif *vif,
 				      struct ieee80211_sta *sta, bool enabled);
-	void (*add_twt_setup)(struct ieee80211_hw *hw,
-			      struct ieee80211_sta *sta,
-			      struct ieee80211_twt_setup *twt);
-	void (*twt_teardown_request)(struct ieee80211_hw *hw,
-				     struct ieee80211_sta *sta,
-				     struct ieee80211_twt_teardown *td);
+	int (*add_twt_setup)(struct ieee80211_hw *hw,
+			     struct ieee80211_sta *sta,
+			     struct ieee80211_twt_setup *twt);
+	int (*twt_teardown_request)(struct ieee80211_hw *hw,
+				    struct ieee80211_sta *sta,
+				    struct ieee80211_twt_teardown *td);
 	int (*add_btwt)(struct ieee80211_hw *hw, struct ieee80211_vif *vif,
 			struct cfg80211_broadcast_twt_settings *btwt);
 	int (*del_btwt)(struct ieee80211_hw *hw, struct ieee80211_vif *vif, u8 id);
diff --git a/net/mac80211/driver-ops.h b/net/mac80211/driver-ops.h
index 0c9385e8..e6457c1d 100644
--- a/net/mac80211/driver-ops.h
+++ b/net/mac80211/driver-ops.h
@@ -1634,42 +1634,49 @@ static inline void drv_sta_set_decap_offload(struct ieee80211_local *local,
 	trace_drv_return_void(local);
 }
 
-static inline void drv_add_twt_setup(struct ieee80211_local *local,
-				     struct ieee80211_sub_if_data *sdata,
-				     struct ieee80211_sta *sta,
-				     struct ieee80211_twt_setup *twt)
+static inline int drv_add_twt_setup(struct ieee80211_local *local,
+				    struct ieee80211_sub_if_data *sdata,
+				    struct ieee80211_sta *sta,
+				    struct ieee80211_twt_setup *twt)
 {
 	struct ieee80211_twt_params *twt_agrt;
+	int ret = -EINVAL;
 
 	might_sleep();
 	lockdep_assert_wiphy(local->hw.wiphy);
 
 	if (!check_sdata_in_driver(sdata))
-		return;
+		return ret;
 
 	twt_agrt = (void *)twt->params;
 
 	trace_drv_add_twt_setup(local, sta, twt, twt_agrt);
-	local->ops->add_twt_setup(&local->hw, sta, twt);
-	trace_drv_return_void(local);
+	ret = local->ops->add_twt_setup(&local->hw, sta, twt);
+	trace_drv_return_int(local, ret);
+
+	return ret;
 }
 
-static inline void drv_twt_teardown_request(struct ieee80211_local *local,
-					    struct ieee80211_sub_if_data *sdata,
-					    struct ieee80211_sta *sta,
-					    struct ieee80211_twt_teardown *td)
+static inline int drv_twt_teardown_request(struct ieee80211_local *local,
+					   struct ieee80211_sub_if_data *sdata,
+					   struct ieee80211_sta *sta,
+					   struct ieee80211_twt_teardown *td)
 {
+	int ret = -EINVAL;
+
 	might_sleep();
 	lockdep_assert_wiphy(local->hw.wiphy);
 	if (!check_sdata_in_driver(sdata))
-		return;
+		return ret;
 
 	if (!local->ops->twt_teardown_request)
-		return;
+		return ret;
 
 	trace_drv_twt_teardown_request(local, sta, td);
-	local->ops->twt_teardown_request(&local->hw, sta, td);
-	trace_drv_return_void(local);
+	ret = local->ops->twt_teardown_request(&local->hw, sta, td);
+	trace_drv_return_int(local, ret);
+
+	return ret;
 }
 
 static inline int drv_add_btwt(struct ieee80211_local *local,
diff --git a/net/mac80211/s1g.c b/net/mac80211/s1g.c
index 01388407..70a4d079 100644
--- a/net/mac80211/s1g.c
+++ b/net/mac80211/s1g.c
@@ -102,24 +102,18 @@ ieee80211_s1g_rx_twt_setup(struct ieee80211_sub_if_data *sdata,
 	struct ieee80211_twt_setup *twt = (void *)mgmt->u.action.u.s1g.variable;
 	struct ieee80211_twt_params *twt_agrt = (void *)twt->params;
 
-	twt_agrt->req_type &= cpu_to_le16(~IEEE80211_TWT_REQTYPE_REQUEST);
+	twt_agrt->req_type &= ~cpu_to_le16(IEEE80211_TWT_REQTYPE_REQUEST);
+
+	/* TWT Information not supported yet */
 	twt->control |= IEEE80211_TWT_CONTROL_RX_DISABLED;
 
-	/* broadcast TWT not supported yet */
-	if (twt->control & IEEE80211_TWT_CONTROL_NEG_TYPE_BROADCAST) {
-		twt_agrt->req_type &=
-			~cpu_to_le16(IEEE80211_TWT_REQTYPE_SETUP_CMD);
+	if (drv_add_twt_setup(sdata->local, sdata, &sta->sta, twt)) {
+		twt_agrt->req_type &= ~cpu_to_le16(IEEE80211_TWT_REQTYPE_SETUP_CMD);
 		twt_agrt->req_type |=
 			le16_encode_bits(TWT_SETUP_CMD_REJECT,
 					 IEEE80211_TWT_REQTYPE_SETUP_CMD);
-		goto out;
 	}
 
-	/* TWT Information not supported yet */
-	twt->control |= IEEE80211_TWT_CONTROL_RX_DISABLED;
-
-	drv_add_twt_setup(sdata->local, sdata, &sta->sta, twt);
-out:
 	ieee80211_s1g_send_twt_setup(sdata, mgmt->sa, sdata->vif.addr, twt);
 }
 
@@ -128,9 +122,9 @@ ieee80211_s1g_rx_twt_teardown(struct ieee80211_sub_if_data *sdata,
 			      struct sta_info *sta, struct sk_buff *skb)
 {
 	struct ieee80211_mgmt *mgmt = (struct ieee80211_mgmt *)skb->data;
+	struct ieee80211_twt_teardown *td = (void *)mgmt->u.action.u.s1g.variable;
 
-	drv_twt_teardown_request(sdata->local, sdata, &sta->sta,
-		(struct ieee80211_twt_teardown *) mgmt->u.action.u.s1g.variable);
+	drv_twt_teardown_request(sdata->local, sdata, &sta->sta, td);
 }
 
 static void
@@ -142,8 +136,14 @@ ieee80211_s1g_tx_twt_setup_fail(struct ieee80211_sub_if_data *sdata,
 	struct ieee80211_twt_params *twt_agrt = (void *)twt->params;
 	struct ieee80211_twt_teardown td;
 
-	td.id = le16_get_bits(twt_agrt->req_type,
-			      IEEE80211_TWT_REQTYPE_FLOWID);
+	if (twt->control & IEEE80211_TWT_CONTROL_NEG_TYPE_BROADCAST) {
+		td.id = le16_get_bits(twt_agrt->bcast.btwt_info,
+				      IEEE80211_BTWT_INFO_ID);
+		td.neg_type = IEEE80211_TWT_NEGO_TYPE_BTWT_MBR_MGMT;
+	} else {
+		td.id = le16_get_bits(twt_agrt->req_type,
+				      IEEE80211_TWT_REQTYPE_FLOWID);
+	}
 
 	drv_twt_teardown_request(sdata->local, sdata, &sta->sta, &td);
 
-- 
2.45.2

